### DISCUSSION TITLE
Test reexecuteQuery on click.

### DISCUSSION THREAD

**Discussion by brenotx** [urql-graphql/urql#3190] [Q&A]
Hi, I want to test a component that renders an error page and has a "try again" button that calls `reexecuteQuery`. 

I want to test the following flow:
 - Components calls `useQuery` but return an error that causes the "try again" button to be rendered.
 - User clicks on the "try again" that calls `reexecuteQuery`
 - the `useQuery` returns valid data and the user can see the data.
 
 Is there a way to test this? I couldn't find any example of how to test a case like this.

**Comment by kitten** [comment]
I’m currently thinking of how we could model an `@urql/testing` library to test cases like these, but I don’t have concrete details on that yet. It may also not directly help you in testing bindings all the time.

That said, we have a couple of examples on what this could look like in our own tests:
- https://github.com/urql-graphql/urql/blob/cb628ca4e9207b1928a5e137fddbd14c33afb6e4/packages/react-urql/src/hooks/useQuery.test.tsx#L189-L195
- https://github.com/urql-graphql/urql/blob/cb628ca4e9207b1928a5e137fddbd14c33afb6e4/packages/react-urql/src/hooks/useQuery.spec.ts#L261-L272
- https://github.com/urql-graphql/urql/blob/cb628ca4e9207b1928a5e137fddbd14c33afb6e4/packages/react-urql/e2e-tests/useQuery.spec.tsx#L224-L228

So, due to us having different unit test setups over time, we actually have three different testing approaches/libraris in place for the React bindings. Generally, they follow a simple pattern however:
- test that the state settles
- click/execute on `useQuery`
- wait for the state to settle / `act` / etc
- check that the loading state is active; or with testing state: test that the new state is applied

So, in short, it really depends on how you test.

Should you test that the `Client` receives the operation? Probably not.
Could you test the `Client` with mock state and that `client.executeQuery` is called on click? Certainly!

So, in short, I'd spy/mock `client.executeQuery`, test that the retry button works after an error and calls it, and then test that the resulting state is applied.

Again, we don't have helpers for this yet, but we do have some advice on how to mock/spy the `Client` here: https://formidable.com/open-source/urql/docs/advanced/testing/#mocking-the-client

**Reply by brenotx** [reply]
Since we are using React Testing Library, we try to create tests that assemble how users interact with your code (component, page, etc.) as much as possible. It would be nice to have a way to write a test case where I can check for text/button/link that will be rendered if my query returns an error, then after clicking on `Try Again`(which calls `reexecuteQuery`) I could wait and query for a new text, possible a data coming from the successful response.

Something  like:
 1. Render the component with the error state. Assert the UI is ok by checking its content.
 2. Click on something that calls `reexecuteQuery` that will return a successful response. Here we don't need to force anything like `act(() => executeQuery());`,  the  `userEvent.click('Try again')` should already trigger the query again.
 3. Assert that the data was rendered properly.
 
 
 Here is what I'm trying to do in my test case:
 
 ```js
 it('should retry to load the data if the user clicks "Try again"', async () => {
  const errorState = {
    executeQuery: jest
      .fn()
      .mockReturnValueOnce(
        fromValue({
          error: new CombinedError({
            networkError: Error('Service unavailable'),
          }),
        })
      )
      .mockReturnValueOnce(fromValue(mockReponse)),
  };

  render(
    <Provider value={errorState}>
      <Students />
    </Provider>
  );

  expect(
    await screen.findByText(
      /Something went wrong/i
    )
  ).toBeInTheDocument();

  expect(errorState.executeQuery).toHaveBeenCalledTimes(1);

  userEvent.click(screen.getByRole('button', { name: /Try again/i }));

  expect(errorState.executeQuery).toHaveBeenCalledTimes(2);

  expect(
    await screen.findByRole('heading', { name: 'Skills' })
  ).toBeInTheDocument();
});
```

The `executeQuery` is called two times but always returns the first response(error response). The test fails on the last `expect` because of that.
